src/descriptor/mod.rs

```diff
@@ -848,7 +978,153 @@ impl<Pk: MiniscriptKey> fmt::Display for Descriptor<Pk> {
+            Descriptor::ShSortedMulti(ref smv) => write!(f, "sh({})", smv),
+            Descriptor::WshSortedMulti(ref smv) => write!(f, "wsh({})", smv),
+            Descriptor::ShWshSortedMulti(ref smv) => write!(f, "sh(wsh({}))", smv),
+        }
+    }
+}
+
+/// Contents of a "sortedmulti" descriptor
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SortedMultiVec<Pk: MiniscriptKey, Ctx: ScriptContext> {
+    /// signatures required
+    pub k: usize,
+    /// public keys inside sorted Multi
+    pub pks: Vec<Pk>,
+    /// The current ScriptContext for sortedmulti
+    pub(crate) phantom: PhantomData<Ctx>,
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn new(k: usize, pks: &[Pk]) -> Result<Self, Error> {
+        // Check the limits before creating a new SortedMultiVec
+        // For example, under p2sh context the scriptlen can only be
+        // upto 520 bytes.
+        let term: miniscript::decode::Terminal<Pk, Ctx> = Terminal::Multi(k, Vec::from(pks));
+        let ms = Miniscript::from_ast(term)?;
+
+        // This would check all the consensus rules for p2sh/p2wsh and
+        // even tapscript in future
+        Ctx::check_ms_validity(&ms)?;
+
+        Ok(Self {
+            k,
+            pks: Vec::from(pks),
+            phantom: PhantomData,
+        })
+    }
+    /// Parse an expression tree into a SortedMultiVec
+    fn from_tree(tree: &Tree) -> Result<Self, Error>
+    where
+        <Pk as FromStr>::Err: ToString,
+    {
+        let k = tree.args[0].name.parse::<usize>().unwrap(); // FIXME: unwrap
+        let pks: Result<Vec<Pk>, _> = tree.args[1..]
+            .iter()
+            .map(|sub| expression::terminal(sub, Pk::from_str))
+            .collect();
+        pks.map(|pks| SortedMultiVec::new(k, &pks))?
+    }
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn translate_pk<FPk, Q, FuncError>(
+        &self,
+        translatefpk: &mut FPk,
+    ) -> Result<SortedMultiVec<Q, Ctx>, FuncError>
+    where
+        FPk: FnMut(&Pk) -> Result<Q, FuncError>,
+        Q: MiniscriptKey,
+    {
+        let pks: Result<Vec<Q>, _> = self.pks.iter().map(&mut *translatefpk).collect();
+        Ok(SortedMultiVec {
+            k: self.k,
+            pks: pks?,
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<Pk: MiniscriptKey + ToPublicKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    /// Create Terminal::Multi containing sorted pubkeys
+    pub fn sorted_node(&self) -> Terminal<bitcoin::PublicKey, Ctx> {
+        let mut pks: Vec<bitcoin::PublicKey> =
+            self.pks.iter().map(|pk| pk.to_public_key()).collect();
+        // Sort pubkeys lexigraphically according to BIP 67
+        pks.sort_by(|a, b| a.to_string().partial_cmp(&b.to_string()).unwrap());
+        Terminal::Multi(self.k, pks)
+    }
+
+    fn encode(&self) -> script::Script {
+        self.sorted_node()
+            .encode(script::Builder::new())
```

Comment:
Here, we can translate the `Pk` to `bitcoin::PublicKeys` as 
` 
        self.sorted_node()
            .translate_pk(
                &mut |pk| Result::Ok::<_, ()>(pk.to_public_key()),
                &mut |_| unreachable!("Sorted multi cannot have Pkh"),
            ). 
.....
`

Reply:
We decided to ignore this comment over IRC